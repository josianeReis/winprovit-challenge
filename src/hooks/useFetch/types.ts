export type Data = Record<any, any>

export type Value = string | number | Date

export type Callback = () => Promise<{
  data: {
    success: boolean
    data: Data[]
    errors?: any[]
    loading: boolean
  }
}>
