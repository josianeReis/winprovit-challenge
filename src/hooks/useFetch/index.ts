import { useCallback, useEffect, useState } from 'react'

import type * as T from './types'

function useFetch(callback: T.Callback) {
  const [data, setData] = useState<any>(undefined)
  const [loading, setLoading] = useState(false)
  const [errors, setErrors] = useState<any>(undefined)

  const refetch = useCallback(
    async () => {
      try {
        setLoading(true)
        const { data } = await callback()
        setData(data)
        setLoading(false)
      } catch (error) {
        console.log('catch catch')
        setErrors(error)
      } finally {
        setLoading(false)
      }
    },
    [callback]
  )

  useEffect(() => {
    refetch()
  }, [refetch])

  return {
    data,
    refetch,
    loading,
    errors,
  }
}
export default useFetch
