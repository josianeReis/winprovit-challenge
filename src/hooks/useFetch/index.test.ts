import { renderHook, act } from '@testing-library/react-hooks'

import useFetch from '.'

describe('useFetch', () => {
  it('Should return users values', async () => {
    const mockArray = Array.from({ length: 10 }).map(() => ({
      id: 2,
      name: "Ervin Howell",
      email: "Shanna@melissa.tv",
      phone: "010-692-6593 x09125",
      username: "Antonette",
      website: "anastasia.net",
      company: {
        bs: "synergize scalable supply-chains",
        catchPhrase: "Proactive didactic contingency",
        name: "Deckow-Crist",
      },
      address: {
        street: "Victor Plains",
        suite: "Suite 879",
        zipcode: "90566-7771",
        city: "Wisokyburgh",
        geo: {
          lat: "-43.9509",
          lng: "-34.4618"
        }
      }
    }))

    const getUsers = () =>
      Promise.resolve({
        data: mockArray,
      })

    // @ts-ignore
    const { result } = await renderHook(() => useFetch(getUsers))

    act(() => {
      expect(result.current).toMatchSnapshot()
    })
  })
})
