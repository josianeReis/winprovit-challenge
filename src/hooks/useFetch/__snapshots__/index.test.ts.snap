// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useFetch Should return users values 1`] = `
{
  "data": [
    {
      "address": {
        "city": "Wisokyburgh",
        "geo": {
          "lat": "-43.9509",
          "lng": "-34.4618",
        },
        "street": "Victor Plains",
        "suite": "Suite 879",
        "zipcode": "90566-7771",
      },
      "company": {
        "bs": "synergize scalable supply-chains",
        "catchPhrase": "Proactive didactic contingency",
        "name": "Deckow-Crist",
      },
      "email": "Shanna@melissa.tv",
      "id": 2,
      "name": "Ervin Howell",
      "phone": "010-692-6593 x09125",
      "username": "Antonette",
      "website": "anastasia.net",
    },
    {
      "address": {
        "city": "Wisokyburgh",
        "geo": {
          "lat": "-43.9509",
          "lng": "-34.4618",
        },
        "street": "Victor Plains",
        "suite": "Suite 879",
        "zipcode": "90566-7771",
      },
      "company": {
        "bs": "synergize scalable supply-chains",
        "catchPhrase": "Proactive didactic contingency",
        "name": "Deckow-Crist",
      },
      "email": "Shanna@melissa.tv",
      "id": 2,
      "name": "Ervin Howell",
      "phone": "010-692-6593 x09125",
      "username": "Antonette",
      "website": "anastasia.net",
    },
    {
      "address": {
        "city": "Wisokyburgh",
        "geo": {
          "lat": "-43.9509",
          "lng": "-34.4618",
        },
        "street": "Victor Plains",
        "suite": "Suite 879",
        "zipcode": "90566-7771",
      },
      "company": {
        "bs": "synergize scalable supply-chains",
        "catchPhrase": "Proactive didactic contingency",
        "name": "Deckow-Crist",
      },
      "email": "Shanna@melissa.tv",
      "id": 2,
      "name": "Ervin Howell",
      "phone": "010-692-6593 x09125",
      "username": "Antonette",
      "website": "anastasia.net",
    },
    {
      "address": {
        "city": "Wisokyburgh",
        "geo": {
          "lat": "-43.9509",
          "lng": "-34.4618",
        },
        "street": "Victor Plains",
        "suite": "Suite 879",
        "zipcode": "90566-7771",
      },
      "company": {
        "bs": "synergize scalable supply-chains",
        "catchPhrase": "Proactive didactic contingency",
        "name": "Deckow-Crist",
      },
      "email": "Shanna@melissa.tv",
      "id": 2,
      "name": "Ervin Howell",
      "phone": "010-692-6593 x09125",
      "username": "Antonette",
      "website": "anastasia.net",
    },
    {
      "address": {
        "city": "Wisokyburgh",
        "geo": {
          "lat": "-43.9509",
          "lng": "-34.4618",
        },
        "street": "Victor Plains",
        "suite": "Suite 879",
        "zipcode": "90566-7771",
      },
      "company": {
        "bs": "synergize scalable supply-chains",
        "catchPhrase": "Proactive didactic contingency",
        "name": "Deckow-Crist",
      },
      "email": "Shanna@melissa.tv",
      "id": 2,
      "name": "Ervin Howell",
      "phone": "010-692-6593 x09125",
      "username": "Antonette",
      "website": "anastasia.net",
    },
    {
      "address": {
        "city": "Wisokyburgh",
        "geo": {
          "lat": "-43.9509",
          "lng": "-34.4618",
        },
        "street": "Victor Plains",
        "suite": "Suite 879",
        "zipcode": "90566-7771",
      },
      "company": {
        "bs": "synergize scalable supply-chains",
        "catchPhrase": "Proactive didactic contingency",
        "name": "Deckow-Crist",
      },
      "email": "Shanna@melissa.tv",
      "id": 2,
      "name": "Ervin Howell",
      "phone": "010-692-6593 x09125",
      "username": "Antonette",
      "website": "anastasia.net",
    },
    {
      "address": {
        "city": "Wisokyburgh",
        "geo": {
          "lat": "-43.9509",
          "lng": "-34.4618",
        },
        "street": "Victor Plains",
        "suite": "Suite 879",
        "zipcode": "90566-7771",
      },
      "company": {
        "bs": "synergize scalable supply-chains",
        "catchPhrase": "Proactive didactic contingency",
        "name": "Deckow-Crist",
      },
      "email": "Shanna@melissa.tv",
      "id": 2,
      "name": "Ervin Howell",
      "phone": "010-692-6593 x09125",
      "username": "Antonette",
      "website": "anastasia.net",
    },
    {
      "address": {
        "city": "Wisokyburgh",
        "geo": {
          "lat": "-43.9509",
          "lng": "-34.4618",
        },
        "street": "Victor Plains",
        "suite": "Suite 879",
        "zipcode": "90566-7771",
      },
      "company": {
        "bs": "synergize scalable supply-chains",
        "catchPhrase": "Proactive didactic contingency",
        "name": "Deckow-Crist",
      },
      "email": "Shanna@melissa.tv",
      "id": 2,
      "name": "Ervin Howell",
      "phone": "010-692-6593 x09125",
      "username": "Antonette",
      "website": "anastasia.net",
    },
    {
      "address": {
        "city": "Wisokyburgh",
        "geo": {
          "lat": "-43.9509",
          "lng": "-34.4618",
        },
        "street": "Victor Plains",
        "suite": "Suite 879",
        "zipcode": "90566-7771",
      },
      "company": {
        "bs": "synergize scalable supply-chains",
        "catchPhrase": "Proactive didactic contingency",
        "name": "Deckow-Crist",
      },
      "email": "Shanna@melissa.tv",
      "id": 2,
      "name": "Ervin Howell",
      "phone": "010-692-6593 x09125",
      "username": "Antonette",
      "website": "anastasia.net",
    },
    {
      "address": {
        "city": "Wisokyburgh",
        "geo": {
          "lat": "-43.9509",
          "lng": "-34.4618",
        },
        "street": "Victor Plains",
        "suite": "Suite 879",
        "zipcode": "90566-7771",
      },
      "company": {
        "bs": "synergize scalable supply-chains",
        "catchPhrase": "Proactive didactic contingency",
        "name": "Deckow-Crist",
      },
      "email": "Shanna@melissa.tv",
      "id": 2,
      "name": "Ervin Howell",
      "phone": "010-692-6593 x09125",
      "username": "Antonette",
      "website": "anastasia.net",
    },
  ],
  "errors": undefined,
  "loading": false,
  "refetch": [Function],
}
`;
